app/parse.py

```diff
@@ -13,10 +47,131 @@ class Product:
     price: float
     rating: int
     num_of_reviews: int
+    # additional_info: dict :TODO test does not cover it
+
+
+PRODUCT_FIELDS = [field.name for field in fields(Product)]
+
+_driver: WebDriver | None = None
+
+
+def get_driver() -> WebDriver:
+    return _driver
+
+
+def set_driver(new_driver: WebDriver) -> None:
+    global _driver
+    _driver = new_driver
+
+
+def parse_hdd_block_prices(
+    product_soup: BeautifulSoup
+) -> dict[str, float] | str:
+    """Functionality works, but the test doesn't cover it"""
+    detailed_url = urljoin(
+        BASE_URL, product_soup.select_one(".title")["href"]
+    )
+    driver = get_driver()
+    driver.get(detailed_url)
+    try:
+        swatches = driver.find_element(By.CLASS_NAME, "swatches")
+    except NoSuchElementException:
+        return "No other HDD"
+
+    buttons = swatches.find_elements(By.TAG_NAME, "button")
+
+    prices = {}
+    for button in buttons:
+        if not button.get_property("disabled"):
+            button.click()
+            prices[button.get_property("value")] = float(driver.find_element(
+                By.CLASS_NAME, "price"
+            ).text.replace("$", ""))
+
+    return prices
+
+
+def parse_rating(product_soup: BeautifulSoup) -> int:
+    try:
+        return int(product_soup.select_one("p[data-rating]")["data-rating"])
+    except TypeError:
+        rating_element = product_soup.select_one(".ratings")
+        star_icons = rating_element.select(".ws-icon.ws-icon-star")
+        return len(star_icons)
+
+
+def parse_single_product(product_soup: BeautifulSoup) -> Product:
+    # hdd_prices = parse_hdd_block_prices(
+    # product_soup
+    # ) :TODO test does not cover it
+    return Product(
+        title=product_soup.select_one(".title")["title"],
+        description=product_soup.select_one(".description").text.replace("\xa0", " "),
+        price=float(product_soup.select_one(".price").text.replace("$", "")),
+        rating=parse_rating(product_soup),
+        num_of_reviews=int(
+            product_soup.select_one(".ratings > p.pull-right").text.split()[0]
+        ),
+        # additional_info={
+        # "hdd_prices": hdd_prices
+        # }, :TODO test does not cover it
+    )
+
+
+def get_single_page_products(page_soup: BeautifulSoup) -> [Product]:
+    products = page_soup.select(".thumbnail")
+
+    return [parse_single_product(product_soup) for product_soup in products]
+
+
+def check_button_more(url: str) -> str:
+    driver = get_driver()
+    driver.get(url)
+    try:
```

Comment:
You can simplify this code instead of usage two try-except:


```
   more_button = driver.find_elements(
        By.CLASS_NAME, "ecomerce-items-scroll-more"
    )

    if more_button:
        while more_button[0].is_displayed():
            more_button[0].click()
            time.sleep(0.1)
```